Plane: Add ability to run aero scripts from any mode

Hwurzburg opened the pull request:
Extends scripting nav_scripting aerobatics rate control to any mode. Current time stamped in nav_scripting structure and checked by attitude and servo(throttle) modules for age to prevent script death from retaining control.

Below is an example of a knife_edge lua script routine with angle programmed by SCR_USER5 and duration by SCR_USER6....SCR_USER1-4 should be set as  2,1,80,20 for height and pitch pid values as in plane_aerobatics.lua example.....set an rc channel to 300 to start and stop knife_edge....mode changes will exit....script stop will exit (25ms keep alive time...ie for 10ms scripts)...arming protection...( I am sure my LUA style is non-optimal...but it works to test this)

caution , do not activate trick while armed and on the ground....same as if you tried to do that manually....you wont be happy

[knife_edge.lua.zip](https://github.com/ArduPilot/ardupilot/files/7877832/knife_edge.lua.zip)


tested along side the previous plane_aerobatics example to test co-existence...no problem (except rolling circles since USER5/6 have be co-opted for the knife-edge....easily fixed by using new local params, but I was lazy since this was just for testing)

open issue: exiting the knife-edge, the yaw jumps ~20 degrees in stabilized modes...haven't been able to figure out what I need to do in the script to fix that....resetting yaw I in C++ code does not fix...

IamPete1 commented:
I think any mode is a bad idea. I would prefer to add support for guided only. The script can then switch to guided mode, do its stuff, and then switch back to your original mode.

tridge commented:
I do want to make this possible in all modes that use the fixed wing rate controllers (but not Q modes). What I'd like to see is an API that enables this feature explicitly for the current mode. Something like this:
 - vehicle:nav_script_enable(MODE_ACRO)

this would enable the feature only for the mode specified, and it would be disabled again as soon as you change modes, so the script would have to enable it again.
The feature would always be enabled when in AUTO in a NAV_SCRIPT_TIME mission item, as per current master

IamPete1 commented:
I really don't like the idea of breaking the rules on manual modes. In master if your in a manual mode you have control. The mode the vehicle is in tells the user what they can expect the vehicle do when they give inputs. With this change the vehicle can do whatever it likes in whatever mode and the user has no way to tell when scripting is taking control. 

My preferred method would be the script looking for the user to switch into guided mode. The script then does its stuff and puts the vehicle back into the last mode when its done. 

Second best would be the script looking for a aux switch, it then switches into guided itself, does its stuff and puts the mode back. 

As implemented this breaks the basic assumptions of what flight modes can do. 

Scripting should absolutely be allowed to do whatever it likes in guided mode, but unleashing it on all modes is asking for trouble. If the script author wants that behaviour they can switch modes from the script, we should not bake it into the API. 

I see scripting as another forum of external control, just like a companion computer, I'm sure you would not suggest we allow "fly to here" from the GCS manual mode. 

I do not see any advantage to allowing this in any mode vs switching to guided for a bit. Adding per mode enable bindings makes it more complex and costs more flash.

Hwurzburg commented:
added the enable as requested....new lua without having to use mode checking for mode changes below:
[knife_edge.lua.zip](https://github.com/ArduPilot/ardupilot/files/7878783/knife_edge.lua.zip)

robustini commented:
@Hwurzburg thanks Henry for this implementation, I find it very useful to be able to run scripts from any mode, it would have been nice to be able to do it in MANUAL too, so the user who uses them will not be naive.
I am trying your script but I don't understand why it gets in the knife but doesn't do any corrections to yaw.
Am I forgetting something?

robustini commented:
Knife edge demo: https://cdn.discordapp.com/attachments/783141293445480470/932534409150758992/knife_script.mp4

Hwurzburg commented:
Script only holds heading not track

tridge commented on ArduPlane/Plane.h:
```diff
@@ -527,8 +527,10 @@ class Plane : public AP_Vehicle {
+        uint32_t current_ms;
         bool done;
     } nav_scripting;
+    bool nav_scripting_enabled;
```
i'd prefer this as an enabled variable inside nav_scripting structure

tridge commented on ArduPlane/commands_logic.cpp:
```diff
@@ -1184,13 +1184,26 @@ void Plane::nav_script_time_done(uint16_t id)
-    if (!nav_scripting_active()) {
-        return false;
-    }
+    if (nav_scripting_enabled) {      
```
indent needs fixing

tridge commented on ArduPlane/Attitude.cpp:
```diff
@@ -495,6 +495,23 @@ void Plane::stabilize()
+            const float rudder = yawController.get_rate_out(nav_scripting.yaw_rate_dps, speed_scaler, false);
+            steering_control.rudder = rudder;
+        }
+        if (AP_HAL::millis() - nav_scripting.current_ms > 25) { //set_target_throttle_rate_rpy has not been called from script in last 100ms
```
also, comment needs to match timeout. I think 50ms would be good

tridge commented on ArduPlane/Attitude.cpp:
```diff
@@ -495,6 +495,23 @@ void Plane::stabilize()
+        }
+        if (AP_HAL::millis() - nav_scripting.current_ms > 25) { //set_target_throttle_rate_rpy has not been called from script in last 100ms
+            nav_scripting.current_ms = 0;
+            yawController.reset_I();
```
why special reset on yaw controller?

tridge commented on ArduPlane/servos.cpp:
```diff
@@ -552,6 +552,10 @@ void Plane::set_servos_controlled(void)
             SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, get_throttle_input(true));
         }
+#if AP_SCRIPTING_ENABLED
+    } else if (plane.nav_scripting.current_ms > 0 && nav_scripting_enabled) {
```
this needs the timeout check, i'd suggest making nav_scripting_active() do the timeout check and ms reset itself, with exception for AUTO in NAV_SCRIPT_TIME

tridge commented on ArduPlane/commands_logic.cpp:
```diff
@@ -1184,13 +1184,26 @@ void Plane::nav_script_time_done(uint16_t id)
+
+void Plane::nav_scripting_enable(uint8_t mode)
+{
+   if (control_mode->mode_number() != mode) {
```
i think we should refuse in Q modes, unless we actually implement the desired rates in Q modes (eg. tailsitters could do backflips)

Hwurzburg commented:
Updated C++ and added scripts for tricks on a switch....this provides a modular approach to having switch activated aerobatics tricks...provided separate Aerobatics examples directories with README...tested extensively in SITL....can co-exist with AUTO nav_script_time missions....can easily change which modes allow scripting override...

will test in a real vehicle as soon as time permits....

Hwurzburg commented:
flight tested and video updated https://youtu.be/XD81TRRgDUM

tridge commented:
I'm very happy to see this functionality, and I think the decision on what modes this should be done in should be made in the lua script, not in the C++ code. Once we have the ability to create new modes inside lua scripts I think we'll move to that, and create modes suitable for the action being performed.
